rfd/000?-database-access.md

```diff
@@ -0,0 +1,512 @@
+  instance) and database users.
+* Auditing of database connections and executed queries.
+
+## Architecture
+
+Architecturally Database Access is very similar to [Application Access](0008-application-access.md).
+Teleport gains a new mode of operation, "database service", similar to "ssh
+service", "application service" or a "kube service".
+
+Database service runs inside the customer private network and proxies database
+client connections received from the Teleport proxy to the target database. It
+is also responsible for protocol parsing of supported databases and authorizing
+requests coming from the proxy (i.e. granting/denying access to particular
+databases/database users).
+
+```
+              |‾‾‾‾‾‾‾|  reverse  |‾‾‾‾‾‾‾‾‾‾‾‾|          |‾‾‾‾‾‾‾‾‾‾|
+psql -------> | proxy | <-------- | db service | -------> | postgres |
+              |_______|   tunnel  |____________|          |__________|
+```
+
+The database client (such as psql) will talk to Teleport web proxy port (`3080`
+by default) which will use multiplexing to detect the database protocol and
+dispatch to an appropriate service.
+
+## Authentication
+
+In this model, there are 3 points where authentication needs to happen.
+
+### Database client <---> proxy
+
+**Mutual TLS.** Database clients, such as psql, will use short-lived x509
+certificates issued by Teleport in order to authenticate with the proxy.
+
+The certificate metadata includes (via extensions) routing information about
+the target database which users log into using `tsh db login` command (see
+below for UX).
+
+### Proxy <---> database service
+
+**Mutual TLS.** The connection that is passed from proxy to a database service
+is upgraded to TLS as well in order to be able to pass identity information
+over the reverse tunnel.
+
+In order to do that, proxy will generate an ephemeral certificate signed
+by the user authority and re-encode all identity and routing information
+extracted from the certificate supplied by the client into it.
+
+### Database service <---> database instance
+
+**Mutual TLS (onprem).**  When connecting to an onprem database instance, the
+database service will use a client certificate for authentication, which it
+will generate on the fly according to the database requirements (for example,
+PostgreSQL requires the database user name to be encoded in the certificate
+as common name).
+
+This also means that the database needs to be configured with Teleport's
+certificate authority and (in general case) use server certificate issued by
+Teleport which can be done using `tctl auth sign` command (see below for UX).
+
+**Password auth (RDS/Aurora).** Amazon RDS/Aurora don't support client cert
+authentication. Instead, they support IAM authentication so Teleport database
+service will generate a short-lived authentication token using RDS API that
+will be used as a password:
+
+https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/UsingWithRDS.IAMDBAuth.html
```

Comment:
I don't like us adding non-standard vendor-specific authn protocols into teleport. It sets the precedent for us to build integrations with users' $favorite_cloud_provider instead of pushing said providers to implement standard APIs. And as we add support for more databases and more custom authn schemes, compatibility matrix starts growing fast.

On the other hand, AWS is a popular platform and we're incentivized to make integration there as easy as possible.

Curious to hear your thoughts on this.

Reply:
It's one of the biggest value adds for customers. I wish AWS and Google settled on a mTLS as a method of auth, but the reality of this if we don't do it, many customers won't use Teleport Database access, so we have to if we want to be competitive in the market :)